pkg/compression/compression.go

```diff
@@ -165,3 +171,57 @@ func AutoDecompress(stream io.Reader) (io.ReadCloser, bool, error) {
 	}
 	return res, decompressor != nil, nil
 }
+
+type estarCompressor struct {
+	file             *os.File
+	writer           io.Writer
+	metadata         map[string]string
+	compressionLevel *int
+}
+
+func (c estarCompressor) Write(p []byte) (int, error) {
+	return c.file.Write(p)
+}
+
+func (c estarCompressor) Close() error {
+	defer c.file.Close()
+	st, err := c.file.Stat()
+	if err != nil {
+		return err
+	}
+	sr := io.NewSectionReader(c.file, 0, st.Size())
+
+	compressionLevel := pgzip.BestCompression
+	if c.compressionLevel != nil {
+		compressionLevel = *c.compressionLevel
+	}
+
+	blob, err := estargz.Build(sr, estargz.WithCompressionLevel(compressionLevel))
+	if err != nil {
+		return err
+	}
+	defer blob.Close()
+
+	c.metadata["containerd.io/snapshot/stargz/toc.digest"] = string(blob.TOCDigest())
+	c.metadata["io.containers.estargz.uncompressed-size"] = fmt.Sprintf("%v", st.Size())
```

Comment:
This needs to be the uncompressed size of the *result* estargz blob (not the source tar).